sunbeam-python/sunbeam/features/instance_recovery/feature.py

```diff
@@ -0,0 +1,168 @@
+# Copyright (c) 2024 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import click
+from packaging.version import Version
+from rich.console import Console
+
+from sunbeam.core.common import BaseStep, run_plan
+from sunbeam.core.deployment import Deployment
+from sunbeam.core.juju import JujuHelper
+from sunbeam.core.manifest import (
+    AddManifestStep,
+    CharmManifest,
+    FeatureConfig,
+    SoftwareConfig,
+)
+from sunbeam.core.terraform import TerraformInitStep
+from sunbeam.features.interface.v1.base import FeatureRequirement
+from sunbeam.features.interface.v1.openstack import (
+    DisableOpenStackApplicationStep,
+    EnableOpenStackApplicationStep,
+    OpenStackControlPlaneFeature,
+    TerraformPlanLocation,
+)
+from sunbeam.steps.hypervisor import ReapplyHypervisorTerraformPlanStep
+from sunbeam.utils import pass_method_obj
+from sunbeam.versions import OPENSTACK_CHANNEL
+
+console = Console()
+
+
+class InstanceRecoveryFeature(OpenStackControlPlaneFeature):
+    version = Version("0.0.1")
+
+    requires = {FeatureRequirement("consul")}
+    name = "instance-recovery"
+    tf_plan_location = TerraformPlanLocation.SUNBEAM_TERRAFORM_REPO
+
+    def default_software_overrides(self) -> SoftwareConfig:
+        """Feature software configuration."""
+        return SoftwareConfig(
+            charms={"masakari-k8s": CharmManifest(channel=OPENSTACK_CHANNEL)}
+        )
+
+    def manifest_attributes_tfvar_map(self) -> dict:
+        """Manifest attributes terraformvars map."""
+        return {
+            self.tfplan: {
+                "charms": {
+                    "masakari-k8s": {
+                        "channel": "masakari-channel",
+                        "revision": "masakari-revision",
+                        "config": "masakari-config",
+                    }
+                }
+            }
+        }
+
+    def pre_enable(self, deployment: Deployment, config: FeatureConfig) -> None:
+        """Handler to perform tasks before enabling the feature."""
+        if self.get_cluster_topology(deployment) == "single":
+            click.echo("WARNING: This feature is not useful in Single node deployment.")
```

Comment:
A bit of editorial nit on this because it's not only with single node that there's an issue. So we should check if we have more than 1 hv and output this warning. (this encompasses single node but also multi-node with only one compute)

`WARNING: This feature is meant to work across multi-hypervisors setup. Current setup detected only has less than 2 hypervisors.`

Reply:
Initially i thought to warn with a check of less than 2 hypervisors but it seems to me documentation is the right place since the user can enable plugin before adding compute nodes. However i did the change for single node because the deployment will have at max 1 compute node only.

Reply:
Ok, then at least:
`"WARNING: This feature is meant for multi-node deployment only." or something, I find the wording a little strange

Reply:
Fair enough, changed as per your suggestion.